### PR TITLE
Modify catch hyperdash behavior to match to stable

### LINKED ISSUE #2627: osu!catch hyperdash behavior is different from stable
- Not pressing the dash button while hyperdashing should give the catcher half speed.
- The target point of the hyperdash should be 1000/60 milliseconds below from the target fruit.
- The hyperdash should end when the catcher reaches the target x-coordinate.

I don't know this is intentional or not but CtB play experience is very sensitive to small differences thus I think the behavior should be precisely matched to the stable.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58783788-osu-catch-hyperdash-behavior-is-different-from-stable?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ekrctb**
Resolves #2627.
I'm not experienced in collaborative works so correct me if I'm doing anything wrong.

### CODE REVIEW

File: osu.Game.Rulesets.Catch.Tests/TestCaseCatcherArea.cs [hunk 1]
```diff
@@ -26,7 +26,6 @@ public class TestCaseCatcherArea : OsuTestCase
         public TestCaseCatcherArea()
         {
             AddSliderStep<float>("CircleSize", 0, 8, 5, createCatcher);
-            AddToggleStep("Hyperdash", t => catcherArea.ToggleHyperDash(t));
```
Comment (ekrctb) [inline osu.Game.Rulesets.Catch.Tests/TestCaseCatcherArea.cs]: I think it is unused? I'm not sure if it fully makes sense to toggle hyper-dash state without specifying speed / targets. So I removed this for now.

**Comment by peppy** [comment]
I don't understand why you changed the bool to no longer be based on the modifier. Could you explain what this refactor gained?

File: osu.Game.Rulesets.Catch/UI/CatcherArea.cs [hunk 2]
```diff
@@ -248,37 +248,46 @@ public bool AttemptCatch(CatchHitObject fruit)
 
                 if (validCatch && fruit.HyperDash)
                 {
-                    HyperDashModifier = Math.Abs(fruit.HyperDashTarget.X - fruit.X) / Math.Abs(fruit.HyperDashTarget.StartTime - fruit.StartTime) / BASE_SPEED;
-                    HyperDashDirection = fruit.HyperDashTarget.X - fruit.X;
+                    var target = fruit.HyperDashTarget;
+                    double timeDifference = target.StartTime - fruit.StartTime;
+                    double positionDifference = target.X * CatchPlayfield.BASE_WIDTH - catcherPosition;
+                    double velocity = positionDifference / Math.Max(1.0, timeDifference - 1000.0 / 60.0);
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/CatcherArea.cs]: what are you referencing to implement this behaviour?

**Comment by ekrctb** [comment]
HyperDashing is not equal to (HyperDashModifier != 1) because A hyperdash can have modifier == 1 (or modifier < 1) on certain conditions.

**Comment by peppy** [comment]
Can you explain those conditions? That sounds like a bug.

File: osu.Game.Rulesets.Catch/UI/CatcherArea.cs (hunk 2 above)
Comment (ekrctb) [reply osu.Game.Rulesets.Catch/UI/CatcherArea.cs]: The leaked osu! source code.

**Comment by ekrctb** [comment]
I can't reproduce the precise condition but I once managed to get hyper speed < 1 on stable and I think manipulating the inequalities show the possibility but my brain is not working good enough to do the math for now.
However, I think the behavior on hyper speed <= 1 can be changed because it rarely happens in practice. Should I change this?

**Comment by peppy** [comment]
As discussed further in discord (i believe), let's change the hyperdash variables back with the assumption that it cannot go beneath a multiplier of 1.0.

**Review by peppy — DISMISSED** [review]

**Review by peppy — APPROVED** [review]

**Pull request merged**